### ISSUE TITLE
Gmail IMAP protocol error

### ISSUE THREAD

**Issue by valgarf** [mailpile/Mailpile#1760]
Mailpile always returns a 'protocol error' and/or 'invalid username or password' message when trying to connect to gmail (imap as well as smtp).

Things I tried:
- I copied the password into the UI multiple times. I am sure it is the correct username/password combination. 
- Other clients can connect to the gmail account. (Imap is activated in gmail and 'insecure' apps are allowed.)
- Mailpile connects to other mail accounts just fine. 

Setup:
- Mailpile is running on a virtual server
- OS: Ubuntu 14.04.5 LTS
- gpg: 1.4.16
- Mailpile version: commit 2f91a9667fc0a6258f50ca529fa6af7c0050e01d

Excerpt from log when trying to download mails:
```
http/log(20): Elapsed: 9.731s (Sending response)
http/log(20): Elapsed: 0.000s (output: Set output mode to: json)
bg/log(50): Traceback (most recent call last):
  File "/opt/Mailpile/mailpile/mail_source/imap.py", line 554, in _connect_imap
    conn = timed(mkconn)
  File "/opt/Mailpile/mailpile/mail_source/imap.py", line 521, in timed
    return RunTimed(timeout, *args, **kwargs)
  File "/opt/Mailpile/mailpile/util.py", line 973, in work
    result.append(func(*args, **kwargs))
  File "/opt/Mailpile/mailpile/mail_source/imap.py", line 553, in mkconn
    int(settings.get('port')))
  File "/usr/lib/python2.7/imaplib.py", line 193, in __init__
    self.welcome = self._get_response()
  File "/usr/lib/python2.7/imaplib.py", line 928, in _get_response
    resp = self._get_line()
  File "/usr/lib/python2.7/imaplib.py", line 1028, in _get_line
    raise self.abort('socket error: EOF')
abort: socket error: EOF

587bd944-06813-5855=ImapMailSource:R An IMAP protocol error occurred
bg/log(50): <ImapMailSource(XXXXX@gmail.com, started 140211478587136)>: Idle: An IMAP protocol error occurred
587bd944-06813-5855=ImapMailSource:R An IMAP protocol error occurred
587be471-0700e-5855=ImapMailSource:R Invalid username or password
bg/log(50): <ImapMailSource(XXXXX@gmail.com, started 140211464161024)>: Idle: Invalid username or password
```

Excerpt when trying to send mail:
```
> http/log(20): Elapsed: 0.014s (Sending response)
> bg/log(50): Traceback (most recent call last):
>   File "/opt/Mailpile/mailpile/mail_source/imap.py", line 554, in _connect_imap
>     conn = timed(mkconn)
>   File "/opt/Mailpile/mailpile/mail_source/imap.py", line 521, in timed
>     return RunTimed(timeout, *args, **kwargs)
>   File "/opt/Mailpile/mailpile/util.py", line 973, in work
>     result.append(func(*args, **kwargs))
>   File "/opt/Mailpile/mailpile/mail_source/imap.py", line 553, in mkconn
>     int(settings.get('port')))
>   File "/usr/lib/python2.7/imaplib.py", line 193, in __init__
>     self.welcome = self._get_response()
>   File "/usr/lib/python2.7/imaplib.py", line 928, in _get_response
>     resp = self._get_line()
>   File "/usr/lib/python2.7/imaplib.py", line 1028, in _get_line
>     raise self.abort('socket error: EOF')
> abort: socket error: EOF
> 
> 587bd944-06813-5855=ImapMailSource:R An IMAP protocol error occurred
> bg/log(50): <ImapMailSource(XXXXX@gmail.com, started 140211478587136)>: Idle: An IMAP protocol error occurred
> 587bd944-06813-5855=ImapMailSource:R An IMAP protocol error occurred
> http/log(20): Elapsed: 1.002s (eventlog: Found 1 events)
> bg/log(20): https://YYYYYYY.ZZ "GET /mailpile/api/0/logs/events/?since=1484516332.204812&gather=1&wait=30 HTTP/1.1" 200 1278
> http/log(20): Elapsed: 1.005s (Sending response)
> http/log(20): Elapsed: 0.000s (output: Set output mode to: json)
> 587be0cf-06d4b-5855=Sendit:R Invalid username or password
> 587be0cf-06d4b-5855=Sendit:i Could not send mail to XXXXX@gmail.com
> Could not send mail to XXXXX@gmail.com
> Traceback (most recent call last):
>   File "/opt/Mailpile/mailpile/plugins/compose.py", line 907, in command
>     events=events)])
>   File "/opt/Mailpile/mailpile/smtp_client.py", line 318, in SendMail
>     RunTimed(300, sm_startup)
>   File "/opt/Mailpile/mailpile/util.py", line 973, in work
>     result.append(func(*args, **kwargs))
>   File "/opt/Mailpile/mailpile/smtp_client.py", line 275, in sm_startup
>     details={'authentication_error': True})
>   File "/opt/Mailpile/mailpile/smtp_client.py", line 168, in fail
>     raise SendMailError(msg, details=details)
> SendMailError: Invalid username or password
> 
> sendmail error: Nothing was sent
> 587bebe8-074b6-5855=EmptyOutbox:c sendmail: Nothing was sent
> Elapsed: 6.344s (sendmail: Nothing was sent)
> 
```

**Comment by BjarniRunar** [comment]
Have you enabled IMAP at the GMail side? And "legacy applications", as prompted during the account setup?

**Comment by valgarf** [comment]
Yes I have enabled IMAP and 'less secure apps'. I just tried with a local installation of Mailpile with the same account and it works fine. I assume gmail blocks connections from my server for other reasons.
My current theory: Even when two-factor authentication is disabled, google requires a second authentication method when logging in from a new PC. I would have to 'unlock' my server first.

In any case: It is most likely not Mailpiles fault and I would not expect Mailpile to jump through all the hoops google uses to discourage third-party applications.
